Remove duplicate ssl setup in sql/qa projects

breskeby opened the pull request:
This is follow up on https://github.com/elastic/elasticsearch/pull/57197 as we run ssl cert certification for each `with-ssl` subproject with identical parameters. This PR introduces an `elasticsearch.test-with-ssl` plugin that configures projects to use pregenerated certificates for testing that are shipped with buildSrc.

- Removes duplicate code from `with-ssl` test projects
- Removes expensive duplicate dynamic creation of ssl certificate generation during precommit run
- Documents how the certificates can be recreated
- Used certs are static and not generated dynamically anymore

Here's a comparison of running `precommit` locally before(https://gradle-enterprise.elastic.co/s/lhoqf5yix5g24/performance/build) and after(https://gradle-enterprise.elastic.co/s/djogrpm6ci7fs/performance/build)

breskeby committed:
Remove duplicate ssl setup in sql/qa projects

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Build)

rjernst reviewed (commented):
One general comment is I wonder why we are still generating these files on the fly. In other projects (eg `:x-pack:qa:smoke-test-plugins-ssl`) we have moved away from generating the ssl support files and instead check them in. What if we were to move them to a shared location in buildSrc and have a plugin that copies them into projects and does the configuration. That would also mean the sql projects here could support fips testing, which I believe is why we moved to static files in the first place.

cc @jkakavas

breskeby committed:
Fix enforcement of task instances

breskeby commented on x-pack/plugin/sql/qa/server/security/with-ssl/build.gradle:
```diff
@@ -148,6 +29,29 @@ integTest.runner {
+  }
+}
+
+tasks.withType(org.elasticsearch.gradle.testclusters.TestClustersAware){
```
good catch. 👀 . Fixed

jkakavas commented:
> One general comment is I wonder why we are still generating these files on the fly. In other projects (eg :x-pack:qa:smoke-test-plugins-ssl) we have moved away from generating the ssl support files and instead check them in. 

The reason is that the SQL client doesn't yet support to be configured with certificate/key material in PEM format and thus we can handle this in the same way we do in other qa projects. That's why we have this : 

```
integTest.runner {
  dependsOn(importClientCertificateInNodeKeyStore)
  onlyIf {
    // Do not attempt to form a cluster in a FIPS JVM, as doing so with a JKS keystore will fail.
    // TODO Revisit this when SQL CLI client can handle key/certificate instead of only Keystores.
    // https://github.com/elastic/elasticsearch/issues/32306
    BuildParams.inFipsJvm == false
  }
}
```

> What if we were to move them to a shared location in buildSrc and have a plugin that copies them into projects and does the configuration. That would also mean the sql projects here could support fips testing, which I believe is why we moved to static files in the first place.

This is a step in the right direction IMHO, but we can always defer that to when we will be able to actually use the key/cert material in the SQL projects

breskeby commented:
> One general comment is I wonder why we are still generating these files on the fly. In other projects (eg `:x-pack:qa:smoke-test-plugins-ssl`) we have moved away from generating the ssl support files and instead check them in. What if we were to move them to a shared location in buildSrc and have a plugin that copies them into projects and does the configuration. That would also mean the sql projects here could support fips testing, which I believe is why we moved to static files in the first place.
> 
> cc @jkakavas

From what I can see there are two dynamic variables in this generation:
1. `san`(SubjectAlternativeName)
2. `dname`(Distinguished Name). 

At least `dname` seems to possible to be made a fixed string I think. for `san` I'm not sure how it is used and I honestly don't know if this can be removed or alternated to be static so we could generate them once and ship as you described. Would be great if we can. as generating those files is like ~3-4s on each execution (and part of `precommit` atm)

breskeby commented:
@jkakavas so I wonder is this dynamic in the creation of those files as described here: https://github.com/elastic/elasticsearch/pull/57319#issuecomment-635800528 required or could we remove that so we can have those files generate once, checked in and reused in all environments (local devs, ci, etc)?

breskeby committed:
Fix imports

jkakavas commented:
No, we don't need to generate those dynamically. We can use catch-all SANs that would work for all 
```
 X509v3 Subject Alternative Name: 
                DNS:localhost, DNS:localhost.localdomain, DNS:localhost4, DNS:localhost4.localdomain4, DNS:localhost6, DNS:localhost6.localdomain6, IP Address:127.0.0.1, IP Address:0:0:0:0:0:0:0:1
```

We have README files in most(wanted to say _all_ but I noticed I didn't put one in `:x-pack:qa:smoke-test-plugins-ssl` ) of the places we use pre-generated keys and certificates describing how these were generated , i.e. : https://github.com/jkakavas/elasticsearch/blob/75a5a64427bed323f76f055050b42d5b05c3a5a7/x-pack/plugin/security/qa/tls-basic/src/test/resources/ssl/README.asciidoc


We also have an issue open to clean these up and do deduplication: https://github.com/elastic/elasticsearch/issues/41646

breskeby commented:
@jkakavas perfect. That should make things simpler and we can remove the overhead on every build run. I look into this.

breskeby committed:
Fix another import

breskeby committed:
Fix configuration refernce

breskeby committed:
Use static data for cert generation

breskeby commented:
@elasticsearchmachine test this please

breskeby committed:
Some fixing in the ssl setup task plumbing

breskeby commented:
@elasticsearchmachine test this please

breskeby committed:
Put certs on test runtime classpath

breskeby committed:
Resolve certs for testing directly

breskeby committed:
Move ssl testing logic into a plugin

breskeby committed:
Fix imports

breskeby committed:
Document test cert creation and cleanup

breskeby committed:
Put more common logic into TestWithSslPlugin

breskeby commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/TestWithSslPlugin.java:
```diff
@@ -0,0 +1,92 @@
+    public void apply(Project project) {
+        File keyStoreDir = new File(project.getBuildDir(), "keystore");
+        TaskProvider<ExportElasticsearchBuildResourcesTask> exportKeyStore = project.getTasks()
+            .register("copyTestCertificates", ExportElasticsearchBuildResourcesTask.class, (t) -> {
```
Alternatively, We could reuse the `buildResources` task. I have no strong opinion but my thinking of using a new task here is to detangle this logic from other uses of  `buildResources` within the elasticsearch build logic.

mark-vieira replied:
Yes. In fact, @rjernst's recent work has _removed_ that global task in favor of individual per-plugin tasks so this is the correct approach going forward.

breskeby committed:
Cleanup

breskeby committed:
Fix checkstyle

breskeby commented:
I've updated the description of this PR and changed the implementation based on @rjernst suggestion.

jkakavas commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/TestWithSslPlugin.java:
```diff
@@ -0,0 +1,89 @@
+            NamedDomainObjectContainer<ElasticsearchCluster> clusters = (NamedDomainObjectContainer<ElasticsearchCluster>) project
+                .getExtensions()
+                .getByName(TestClustersPlugin.EXTENSION_NAME);
+            clusters.all(c -> {
```
Could we also have a conditional here that if we are `inFipsJvm()` it would use key and certificate PEM files to set up TLS instead of keystores so that it can work even when the subproject that uses this plugin is running tests in fips mode ? i.e.
 
```
*.ssl.key
*.ssl.secure_key_passphrase
*.ssl.certificate
``` 

instead of 

```
*.ssl.keystore.path
*.ssl.keystore.secure_password
```

and 

```
*.ssl.certificate_authorities
```

instead of 

```
*.ssl.keystore.path
```
or
```
*.ssl.truststore.path
```

breskeby replied:
@jkakavas same as above. Would like to tackle this in a separate story

jkakavas replied:
I'm fine with that as long as we don't apply this plugin to any other subproject until we have the fips handling because it will start failing in CI. ( sql is fine because we mute the tests in fips eitherway )

jkakavas reviewed (commented):
I love this, it will help us untangle much of the setup in the security tests. I added some comments and questions on whether we can make this plugin slightly more generic so that it can handle other test use cases

breskeby committed:
Apply feedback: Rename *.cert to *.crt files

breskeby commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/TestWithSslPlugin.java:
```diff
@@ -0,0 +1,89 @@
+                .getByName(TestClustersPlugin.EXTENSION_NAME);
+            clusters.all(c -> {
+                // ceremony to set up ssl
+                c.setting("xpack.security.transport.ssl.keystore.path", "test-node.jks");
```
yes, you can override them in the build script after applying the plugin

breskeby commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/TestWithSslPlugin.java:
```diff
@@ -0,0 +1,89 @@
+        File keyStoreDir = new File(project.getBuildDir(), "keystore");
+        TaskProvider<ExportElasticsearchBuildResourcesTask> exportKeyStore = project.getTasks()
+            .register("copyTestCertificates", ExportElasticsearchBuildResourcesTask.class, (t) -> {
+                t.copy("test/ssl/test-client.cert");
```
done

breskeby commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/TestWithSslPlugin.java:
```diff
@@ -0,0 +1,89 @@
+                t.copy("test/ssl/test-client.cert");
+                t.copy("test/ssl/test-client.jks");
+                t.copy("test/ssl/test-node.cert");
+                t.copy("test/ssl/test-node.jks");
```
@jkakavas I think we should be able to do that. Can you raise a separate issue for this? I'd like to keep this PR small and focussed.

jkakavas replied:
raised https://github.com/elastic/elasticsearch/issues/57606

breskeby committed:
Apply feedback regarding generated test certs

- make them valid for 10 years
- Add note to readme about date of expiration

breskeby commented on buildSrc/src/main/resources/test/ssl/README.md:
```diff
@@ -0,0 +1,19 @@
+The certificates are generated using catch-all SAN in the following procedure:
+
+1. Generate the node's keystore:
+   `keytool -genkey -alias test-node -keystore test-node.jks -keyalg RSA -keysize 2048 -validity 712 -dname CN="Elasticsearch Build Test Infrastructure" -keypass keypass -storepass keypass -ext san=dns:localhost,dns:localhost.localdomain,dns:localhost4,dns:localhost4.localdomain4,dns:localhost6,dns:localhost6.localdomain6,ip:127.0.0.1,ip:0:0:0:0:0:0:0:1`
```
done

jkakavas commented on buildSrc/build.gradle:
```diff
@@ -222,6 +222,8 @@ if (project != rootProject) {
     exclude '**/*.wav'
     exclude '**/*.p12'
+    exclude '**/*.jks'
+    exclude '**/*.cert'
```
```suggestion
    exclude '**/*.crt'
```

breskeby replied:
fixed

breskeby committed:
Fix forbiddenPattern task after cert -> crt renaming

breskeby committed:
Another fix after cert -> crt renaming

breskeby commented:
@elasticmachine test this please

mark-vieira commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/TestWithSslPlugin.java:
```diff
@@ -0,0 +1,89 @@
+                t.setOutputDir(keyStoreDir);
+            });
+
+        project.getPlugins().withType(StandaloneRestTestPlugin.class).configureEach(restTestPlugin -> {
```
How does this compare to `project.getPluginManager().withPlugin()`?

mark-vieira commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/TestWithSslPlugin.java:
```diff
@@ -0,0 +1,89 @@
+
+        project.getPlugins().withType(StandaloneRestTestPlugin.class).configureEach(restTestPlugin -> {
+            SourceSetContainer sourceSets = project.getExtensions().getByType(SourceSetContainer.class);
+            SourceSet testSourceSet = sourceSets.getByName("test");
```
We can use `Util.getJavaTestSourceSet()`.

breskeby replied:
done

mark-vieira commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/TestWithSslPlugin.java:
```diff
@@ -0,0 +1,89 @@
+            testSourceSet.compiledBy(exportKeyStore);
+
+            project.getTasks()
+                .withType(org.elasticsearch.gradle.testclusters.TestClustersAware.class)
```
Do we have to use the fully qualified name here? Can we add an import?

breskeby replied:
done

mark-vieira commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/TestWithSslPlugin.java:
```diff
@@ -0,0 +1,89 @@
+
+            // Tell the tests we're running with ssl enabled
+            project.getTasks()
+                .withType(RestIntegTestTask.class)
```
We should instead configure all tasks of `RestTestRunnerTask` type. There can be instances where we create these directly w/o creating a `RestIntegTestTask`. Also, the latter will hopefully dissapear soon.

mark-vieira replied:
Also, would we not want this for unit tests as well? Or is this very specific to integration tests?

breskeby replied:
changed to use `RestTestRunnerTask`. So far I don't think we have unit tests for this kind of testing. I don't see how this would look like tbh.

mark-vieira reviewed (approved):
Man this is about 1000% better. 👍

breskeby committed:
Fix generated certificates

breskeby committed:
Apply review feedback

- Minor cleanup
